controllers/nonadmin_controller.go

```diff
@@ -0,0 +1,205 @@
+	nonAdminImage := r.getNonAdminImage(dpa)
+	if len(nonAdminImage) == 0 {
+		return fmt.Errorf("no Non Admin Controller image found in RELATED_IMAGE_NON_ADMIN_CONTROLLER environment variable or unsupportedOverrides")
+	}
+	ensureRequiredLabels(deploymentObject)
+	ensureRequiredSpecs(deploymentObject, nonAdminImage)
+	return nil
+}
+
+func ensureRequiredLabels(deploymentObject *appsv1.Deployment) {
+	maps.Copy(deploymentLabels, controlPlaneLabel)
+	deploymentObjectLabels := deploymentObject.GetLabels()
+	if deploymentObjectLabels == nil {
+		deploymentObject.SetLabels(deploymentLabels)
+	} else {
+		for key, value := range deploymentLabels {
+			deploymentObjectLabels[key] = value
+		}
+		deploymentObject.SetLabels(deploymentObjectLabels)
+	}
+}
+
+func ensureRequiredSpecs(deploymentObject *appsv1.Deployment, image string) {
+	namespaceEnvVar := corev1.EnvVar{
+		Name:  "WATCH_NAMESPACE",
+		Value: deploymentObject.Namespace,
+	}
+
+	deploymentObject.Spec.Replicas = pointer.Int32(1)
+	deploymentObject.Spec.Selector = &metav1.LabelSelector{
+		MatchLabels: controlPlaneLabel,
+	}
+	templateObjectLabels := deploymentObject.Spec.Template.GetLabels()
+	if templateObjectLabels == nil {
+		deploymentObject.Spec.Template.SetLabels(controlPlaneLabel)
+	} else {
+		templateObjectLabels[controlPlaneKey] = controlPlaneLabel[controlPlaneKey]
+		deploymentObject.Spec.Template.SetLabels(templateObjectLabels)
+	}
+	if len(deploymentObject.Spec.Template.Spec.Containers) == 0 {
+		deploymentObject.Spec.Template.Spec.Containers = []corev1.Container{{
+			Name:            nonAdminObjectName,
+			Image:           image,
+			ImagePullPolicy: corev1.PullAlways,
+			Env:             []corev1.EnvVar{namespaceEnvVar},
+		}}
+	} else {
+		for _, container := range deploymentObject.Spec.Template.Spec.Containers {
+			if container.Name == nonAdminObjectName {
+				container.Image = image
+				container.ImagePullPolicy = corev1.PullAlways
+				container.Env = []corev1.EnvVar{namespaceEnvVar}
+				break
+			}
+		}
+	}
+	deploymentObject.Spec.Template.Spec.RestartPolicy = corev1.RestartPolicyAlways
+	deploymentObject.Spec.Template.Spec.ServiceAccountName = nonAdminObjectName
+}
+
+func (r *DPAReconciler) checkNonAdminEnabled(dpa *oadpv1alpha1.DataProtectionApplication) bool {
+	// TODO https://github.com/openshift/oadp-operator/pull/1316
+	if dpa.Spec.Features != nil &&
+		dpa.Spec.Features.NonAdmin != nil &&
+		dpa.Spec.Features.NonAdmin.Enable != nil {
+		return *dpa.Spec.Features.NonAdmin.Enable
+	}
+
+	return false
+}
+
+func (r *DPAReconciler) getNonAdminImage(dpa *oadpv1alpha1.DataProtectionApplication) string {
+	// TODO https://github.com/openshift/oadp-operator/pull/1316
+	unsupportedOverride := dpa.Spec.UnsupportedOverrides[oadpv1alpha1.NonAdminControllerImageKey]
+	if unsupportedOverride != "" {
+		return unsupportedOverride
+	}
+
+	environmentVariable := os.Getenv("RELATED_IMAGE_NON_ADMIN_CONTROLLER")
+	return environmentVariable
```

Comment:
Need to return a valid and non-empty default image url value here so that non-admin controller deployment does not fail. We can return master/latest build image. For downstream case, the env var will always exist on CSV. But we need to handle the case when `unsupportedOverrides` is not used and not error out here.

Reply:
what is the difference between downstream and development? env var will always exist on CSV here in the development as well, no?

Reply:
- Downstream and upstream CSVs are different, the RELATED_* images env var values are different
- We are just thinking of worst case scenario and returning a valid image I think, also in view of keeping things uniform, we do similar stuff for velero deployment. ref: https://github.com/openshift/oadp-operator/blob/e37936618208aace4feaf77ada603bc2ef184bf4/controllers/velero.go#L514

Reply:
done